spec/formatting.md

```diff
@@ -24,6 +33,15 @@ Formatting of a _message_ is defined by the following operations:
   The resolution of _text_ is rather straightforward,
   and is detailed under _literal resolution_.
 
+Implementations are not required to expose
+the _expression resolution_ and _pattern selection_ operations to their users,
+or even use them in their internal processing,
+as long as the final _formatting_ result is made available to users
+and the observable behavior of the _formatting_ matches that described here.
+
+_Attributes_ MUST NOT have any effect on the formatted output of a _message_,
+or be made available to function implementations.
```

Comment:
As suggested by @macchiati in https://github.com/unicode-org/message-format-wg/pull/879#discussion_r1749510245, but with lower-case "nor" as it's not an RFC 2119 term:
```suggestion
_Attributes_ MUST NOT have any effect on the formatted output of a _message_,
nor be made available to function implementations.
```